rfcs/20200929-keras-mixed-precision.md

```diff
@@ -0,0 +1,1182 @@
+* `LossScaleOptimizer` all-reduces in float16 instead of float32. This will add some redundant casts and operations that grappler will remove in graph mode. This will probably not be implemented by TF 2.4. [Link](#all-reducing-in-float16).
+* The function `tf.keras.mixed_precision.experimental.get_layer_policy` is being removed. The layer policy instead is accessible with the property `Layer.dtype_policy`
+
+## Non-breaking changes
+
+* The policy constructor argument `loss_scale` defaults to the string "auto" instead of a private object. The default value still does the same thing: use a dynamic loss scale for the "mixed_float16" policy and no loss scale for other policies. [Link](#dtype-policies-overview).
+* A warning is issued when the "mixed_bfloat16" policy is used on hardware that does not support it. This will likely not be implemented by TF 2.4. [Link](#the-global-policy).
+* The string "mixed_float16" or "mixed_bfloat16" can be passed to the `dtype` parameter of a layer constructor. [Link](#layers).
+* Layers now have properties `dtype_policy`, `compute_dtype`, and `variable_dtype`. [Link](#layers).
+* A warning is issued if a `tf.Variable` is assigned as an attribute of a mixed precision layer, as doing so will not wrap the variable with an AutoCastVaraible. [Link](#layer-variables).
+* Cross entropy will be fused with softmax in more cases. [Link](#softmax-and-cross-entropy).
+* A warning is issued if cross entropy gets a float16 input and does not fuse with a preceding softmax. [Link](#softmax-and-cross-entropy).
+* LossScaleOptimizer’s `loss_scale` argument defaults to "dynamic". [Link](#lossscaleoptimizer-overview).
+* `LossScaleOptimizer` exposes the `inner_optimizer` attribute. [Link](#lossscaleoptimizer-api).
+* `LossScaleOptimizer` delegates `__getattribute__` to the wrapped optimizer if the attribute is a hyperparameter. [Link](#lossscaleoptimizer-__getattribute__-and-__setattr__-delegation).
+* `LossScaleOptimizer` delegates `_setattr__` to the wrapped optimizer if the attribute is a hyperparameter. [Link](#lossscaleoptimizer-__getattribute__-and-__setattr__-delegation).
+* [RFC #234](https://github.com/tensorflow/community/pull/234), titled "Easily Customizable Optimizer.minimize", is implemented. This is not part of the mixed precision API but has some features that make loss scaling easier to use. In particular, it allows a tensor instead of a function to be passed to `Optimizer.minimize`, making `minimize` easier to use. `LossScaleOptimizer.minimize` is recommended over `LossScaleOptimizer.apply_gradients`, as the former automatically scales the loss and gradients while the latter requires the user to manually scale the loss and gradients.
+
+# Alternatives Considered
+
+## Op-based autocasting API
```

Comment:
Op-based autocasting with down-graph type inheritance is important for providing robust accuracy in mixed precision training. This is particularly true in experimental and research applications. For example, early implementations of batch- and layer-norm used a sequence of low-level operations such as the following.
 
bn(x) = y * (x-mean(x)) / sqrt(var(x)+epsilon) + beta
 
Without computing the mean and variance ops in fp32, training will either diverge or achieve reduced accuracy for many important networks such as InceptionV3, ResNet50, and Xception.
 
Of course, once the utility of an experimental layer is proven, it may be replaced with a “fused” custom kernel that provides optimized performance and handles intermediate precisions internally. However, relatively few layers will merit this effort. Novel model architectures are an important area of innovation, and mixed precision needs to be robust so that researches and early adopters can benefit from the full performance of modern architectures.